### PR TITLE
Added service client to fetch hardware-related parameters

### PR DESCRIPTION
**Pull request by JohanEThomsen**
Added service client to fetch parameters related to the hardware (model+version of cellular module and iccid of SIM)

### CODE REVIEW

File: src/api/hardware.toit [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import system.services
+import ..signal show SignalQuality
+
+interface HardwareService:
+  static SELECTOR ::= services.ServiceSelector
+      --uuid="3909a751-0a14-4409-acf3-7310920bb42a"
+      --major=0
+      --minor=1
+
+  iccid -> string?
```
Comment (kasperl) [inline src/api/hardware.toit:14]: I am leaning towards just including this in the core SDK service.

File: src/api/hardware.toit [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+class HardwareServiceClient extends services.ServiceClient implements HardwareService:
+  static SELECTOR ::= HardwareService.SELECTOR
+  constructor selector/services.ServiceSelector=SELECTOR:
+    assert: selector.matches SELECTOR
+    super selector
+
+  iccid -> string?:
+    result := invoke_ HardwareService.ICCID_INDEX null
+    return result ? result : null
```
Comment (kasperl) [inline src/api/hardware.toit:31]:
```suggestion
    return invoke_ HardwareService.ICCID_INDEX null
```

File: src/api/hardware.toit [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+  model -> string?:
+    result := invoke_ HardwareService.MODEL_INDEX null
+    return result ? result : null
```
Comment (kasperl) [inline src/api/hardware.toit:35]:
```suggestion
    return invoke_ HardwareService.MODEL_INDEX null
```

File: src/api/hardware.toit [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+  version -> string?:
+    result := invoke_ HardwareService.VERSION_INDEX null
+    return result ? result : null
```
Comment (kasperl) [inline src/api/hardware.toit:39]:
```suggestion
    return invoke_ HardwareService.VERSION_INDEX null
```

File: src/api/signal.toit [hunk 5]
```diff
@@ -22,4 +22,4 @@ class SignalServiceClient extends services.ServiceClient implements SignalServic
 
   quality -> SignalQuality?:
     result := invoke_ SignalService.QUALITY_INDEX null
-    return result ? (SignalQuality --power=result[0] --quality=result[1]) : null
+    return result ? (SignalQuality --power=result[0] --quality=result[1]) : null
```
Comment (kasperl) [inline src/api/signal.toit:25]: What changed here? Just the newline at the end? We always have a trailing newline, so you should probably revert this part.

File: src/base/service.toit [hunk 6]
```diff
@@ -240,3 +242,28 @@ class SignalServiceHandler_ implements ServiceHandler SignalService:
     if not driver: return null
     result := driver.signal_quality
     return result ? [ result.power, result.quality ] : null
+
+class HardwareServiceHandler_ implements ServiceHandler HardwareService:
```
Comment (kasperl) [inline src/base/service.toit:246]: The argument for keeping this separate from the cellular service is probably that these make sense even when you're not connected yet.

Hmm. Maybe we should combine `SignalService` and `HardwareService` under some reasonable name.

File: src/hardware.toit [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+model -> string?:
+  service := service_
+  if not service: throw "cellular unavailable"
+  return service.model
+
+version -> string?:
+  service := service_
+  if not service: throw "cellular unavailable"
+  return service.version
```
Comment (kasperl) [inline src/hardware.toit:23]: Add newline at end of file.

**Review by kasperl — COMMENTED** [review]
Good stuff. Here are some comments:

File: src/api/hardware.toit (hunk 1 above)
Comment (JohanEThomsen) [reply src/api/hardware.toit:14]: Fine by me!

File: src/api/signal.toit (hunk 5 above)
Comment (JohanEThomsen) [reply src/api/signal.toit:25]: Yeah that was a mistake on my part. I'll revert it

File: src/base/service.toit (hunk 6 above)
Comment (JohanEThomsen) [reply src/base/service.toit:246]: Yeah that would make sense. They are very small, isolated services right now. 
Maybe we could create a combined `CellularStateService`?

File: src/api/hardware.toit (hunk 1 above)
Comment (kasperl) [reply src/api/hardware.toit:14]: Cool. I think it makes sense to land this as is. If the core SDK service APIs end up supporting this, we can always deprecate the HardwareService classes later -- and they do not hurt at all as is.

File: src/base/service.toit (hunk 6 above)
Comment (kasperl) [reply src/base/service.toit:246]: I like that. Feel free to do it in another PR or fix this one :)

Comment (JohanEThomsen) [reply src/base/service.toit:246]: I justed threw in a new commit combining the two :-)

File: src/base/service.toit [hunk 8]
```diff
@@ -80,6 +81,7 @@ abstract class CellularServiceProvider extends ProxyingNetworkServiceProvider:
         --tags=["cellular"]
     provides CELLULAR_SELECTOR --handler=this
     provides SignalService.SELECTOR --handler=(SignalServiceHandler_ this)
+    provides HardwareService.SELECTOR --handler=(HardwareServiceHandler_ this)
```
Comment (kasperl) [inline src/base/service.toit:84]: This looks a bit broken now.

Comment (JohanEThomsen) [reply src/base/service.toit]: Yup I was too fast. Should be fixed now

**Review by kasperl — APPROVED** [review]
Looks great!

**Pull request merged**